src/Matterhorn/State/Attachments.hs

```diff
@@ -54,3 +62,50 @@ showAttachmentFileBrowser = do
     browser <- liftIO $ Just <$> FB.newFileBrowser FB.selectNonDirectories (AttachmentFileBrowser tId) filePath
     csCurrentTeam.tsEditState.cedFileBrowser .= browser
     setMode ManageAttachmentsBrowseFiles
+
+attachFileByPath :: Text -> MH ()
+attachFileByPath txtPath = do
+    let strPath = unpack txtPath
+    fileInfo <- liftIO $ FB.getFileInfo strPath strPath
+    case FB.fileInfoFileStatus fileInfo of
+        Left _ -> do
+            mhLog LogError $ "Failed to obtain FileInfo for file: " <> txtPath
+            mhError $ BadFile txtPath
+        Right _ -> tryAddAttachment [fileInfo]
+
+tryAddAttachment :: [FB.FileInfo] -> MH ()
+tryAddAttachment entries = do
+    forM_ entries $ \entry -> do
+        -- Is the entry already present? If so, ignore the selection.
+        es <- use (csCurrentTeam.tsEditState.cedAttachmentList.L.listElementsL)
+        let matches = (== FB.fileInfoFilePath entry) .
+                          FB.fileInfoFilePath .
+                          attachmentDataFileInfo
+        case Vector.find matches es of
+            Just _ -> return ()
+            Nothing -> do
+                tryReadAttachment entry >>= \case
+                    Right a -> do
+                        oldIdx <- use (csCurrentTeam.tsEditState.cedAttachmentList.L.listSelectedL)
+                        let newIdx = if Vector.null es
+                                     then Just 0
+                                     else oldIdx
+                        csCurrentTeam.tsEditState.cedAttachmentList %= L.listReplace (Vector.snoc es a) newIdx
```

Comment:
This right here is the bit that needs to be run whenever an attachment is added. `attachFileByPath` doesn't do this, and it's the key part of the state change that makes it so the attachment gets sent to the server. The `cedAttachmentList` is what is used to build the message when it is sent.

Reply:
(by "build the message" I mean "include the attachments when building the message")

Reply:
To see the usage of this, see `Matterhorn.State.handleInputSubmission`.

Reply:
I'm not sure I understand completely.  It looks to me like this code does get called, provided no errors arise.  `attachFileByPath` ultimately delegates the work to `tryAddAttachment` (which is also the function called by the attachment browsing code).

Locally, I used a bunch of logging commands to debug this.  I tried outputting the length of the attachment list vector before and after the command is executed, and it seemed to be working (moving from 0 to 1).

Reply:
Oh, I was mistaken - I didn't see the call to `tryAddAttachment` in `attachFileByPath`. Hmm.